src/src/ChaiBuffer.hpp

```diff
@@ -0,0 +1,176 @@
+/*
```

Comment:
`ChaiVector` had a lot of functionality that wasn't unique to `CHAI` such as a `push_back` and `insert`. The `ChaiBuffer` and `StackBuffer` classes just deal with allocation and whatnot and the `bufferManipulation` namespace implements all the additional fancy semantics. This way we can add new backers for the `Array` class easily by making a `Buffer` class that contains not much more than a `reallocate` and `free` method.

In addition it will no longer require the friendship relation between `ChaiVector` and basically every `Array` class that contained it.